### PR TITLE
fix #39556 (infer_freq not working with freq="H" and DST

### LINKED ISSUE #39556: BUG: 'infer_freq' does not work with tz != "UTC"
The issue has been reported in #8772 which has been closed asking for an example (https://github.com/pandas-dev/pandas/issues/8772#issuecomment-523820768).

Here is the example
```python

import pandas

for tz in [None, "UTC", "CET"]:
    index = pandas.date_range("2018", "2019", tz=tz, closed="left", freq='H')
    print(f"tz = {tz}, index freq = {index.freq}, inferred index freq = {pandas.infer_freq(index)}")

```

#### Problem description

`pandas.infer_freq` can infer the frequency when the index has no tz or tz=="UTC". But for tz=="CET", it returns None.

This is due to a special handling in https://github.com/pandas-dev/pandas/blob/master/pandas/tseries/frequencies.py#L198 to cater for timezone DST in frequencies beyond the hour (day, month,...). However for regular hourly frequencies, the delta which is constant when considering real time/UTC will now show multiple deltas (0 for the DST in october, 2H for the DST in march, 1H for the rest).
Probably needs two paths, one for business frequencies (days, months, ...) and one for hour and below frequencies. Not sure about frequencies between H and D (like 6H) but probably to be considered as business frequencies.

#### Expected Output

```
# output
# tz = None, index freq = <Hour>, inferred index freq = H
# tz = UTC, index freq = <Hour>, inferred index freq = H
# tz = CET, index freq = <Hour>, inferred index freq = None

```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : b5958ee1999e9aead1938c0bba2b674378807b3d
python           : 3.8.5.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.18362
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : French_Belgium.1252

pandas           : 1.1.5
numpy            : 1.19.4
pytz             : 2020.4
dateutil         : 2.8.1
pip              : 20.2.4
setuptools       : 50.3.2
Cython           : 0.29.17
pytest           : 6.1.2
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.2
html5lib         : None
pymysql          : None
psycopg2         : 2.8.6 (dt dec pq3 ext lo64)
jinja2           : 2.11.2
IPython          : 7.19.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : None
fsspec           : 0.8.4
fastparquet      : 0.4.2
gcsfs            : None
matplotlib       : 3.3.4
numexpr          : 2.7.2
odfpy            : None
openpyxl         : 3.0.5
pandas_gbq       : None
pyarrow          : 2.0.0
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.4
sqlalchemy       : None
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : 0.52.0

</details>

### PR DESCRIPTION
**Pull request by sdementen**
- check that the delta are unique before checking if they are day multiples
- add test with freq="H" that raises the bug

- [x] closes #39556 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry => not sure where to enter this ...

### CODE REVIEW

**Review by arw2019 — COMMENTED** [review]
Thanks @sdementen for the PR!

we'll need a whatsnew (I expect targeting 1.3)

File: pandas/tseries/frequencies.py [hunk 1]
```diff
@@ -239,17 +239,18 @@ def get_freq(self) -> Optional[str]:
...
 
         # Business hourly, maybe. 17: one day / 65: one weekend
         if self.hour_deltas in ([1, 17], [1, 65], [1, 17, 65]):
             return "BH"
+
         # Possibly intraday frequency.  Here we use the
         # original .asi8 values as the modified values
         # will not work around DST transitions.  See #8772
-        elif not self.is_unique_asi8:
+        if not self.is_unique_asi8:
```
Comment (arw2019) [inline pandas/tseries/frequencies.py:253]: unnecessary diff tho ofc doesn't change the logic

Comment (sdementen) [reply pandas/tseries/frequencies.py:253]: While the code does the same with or without the diff, the new version is more readable/understandable in terms of logic (as the first if has a return, it cuts the flow of the function and the elif is just equivalent to a new if ... hmmm, not sure to make my point clear ;-) )

Comment (arw2019) [reply pandas/tseries/frequencies.py:253]: yeah that's fine

File: pandas/tseries/frequencies.py [hunk 2]
```diff
@@ -239,17 +239,18 @@ def get_freq(self) -> Optional[str]:
         if not self.is_monotonic or not self.index._is_unique:
             return None
 
-        delta = self.deltas[0]
-        if _is_multiple(delta, _ONE_DAY):
+        delta = min(self.deltas)
```
Comment (jbrockmendel) [inline pandas/tseries/frequencies.py:242]: i think the idea behind using deltas[0] was that deltas should be unique at this point.  is that not the case?

Comment (sdementen) [reply pandas/tseries/frequencies.py:242]: It is not unique when the index has a business day frequency as you have delays of 1 day or 3 days (for the weekend).
A first version of the bugfix I tried was to check first the unicity and then take deltas[0] to fix the issue with freq=H and DST but it broke the test for business days

**Comment by sdementen** [comment]
I have added the whatsnew entry

File: pandas/tseries/frequencies.py (hunk 2 above)
Comment (sdementen) [reply pandas/tseries/frequencies.py]: after reading through the doc of unique_deltas I see the self.deltas area already sorted => no need to take the min.
the key is to check that delta !=0 as with freq="H" and tz with DST, the minimum delta is 0 (in local time)

**Comment by sdementen** [comment]
There is some check that fails after having added the what's new entry. Any clue why?

**Comment by lithomas1** [comment]
@sdementen The CI failures are due to #39688. They should be gone if you merge master.

**Comment by jbrockmendel** [comment]
cc @mroeschke i think this is an hour-based analogue of the problem of freq=Day vs freq=DayDST.

i.e. this fixes one problem but will introduce others.  im hesitant to do that, but the long-term fix has been stuck in limbo for a while

**Comment by sdementen** [comment]
> @sdementen The CI failures are due to #39688. They should be gone if you merge master.

Do I need to do this myself? Or will this automatically be solved once someone merge my PR into master?

**Comment by arw2019** [comment]
> > @sdementen The CI failures are due to #39688. They should be gone if you merge master.
> 
> Do I need to do this myself? Or will this automatically be solved once someone merge my PR into master?

you wanna merge master yourself yeah

**Comment by mroeschke** [comment]
I haven't used `infer_freq` much, but if `infer_freq` = "infer the wall time frequency" then yes this the same issue as freq=Day vs freq=DayDST.

**Comment by sdementen** [comment]
merged with master, all tests pass, ready to merge ;-)

**Comment by sdementen** [comment]
@arw2019 @mroeschke , do I need to do something more re this PR for it to be merged ?

**Review by jreback — CHANGES_REQUESTED** [review]
can u also add an explicit test that is similar to the OP

**Comment by sdementen** [comment]
> can u also add an explicit test that is similar to the OP

The test `test_infer_freq_tz_transition` is similar to my OP in the issue. The bug was not revealing itself because the frequency "H" was not used (only "3H") and it is this frequency "H" that I uses in my OP.

File: pandas/tests/tseries/frequencies/test_inference.py [hunk 3]
```diff
@@ -267,7 +267,7 @@ def test_infer_freq_tz(tz_naive_fixture, expected, dates):
     ],
 )
 @pytest.mark.parametrize(
```
Comment (jreback) [inline pandas/tests/tseries/frequencies/test_inference.py:269]: this is not exactly the same as the OP (though it may have revelead the issue). this is a naive fixture (IOW the OP worked for naive & UTC) ,but NOT for other tzs.

so can also make this more comprehensive

Comment (sdementen) [reply pandas/tests/tseries/frequencies/test_inference.py:269]: Not sure to get your comment yet ...

When I run the test `test_infer_freq_tz_transition`, it runs for a lot of tzs (None, UTC, US/Eastern, Asia/Tokyo, ...), for date_pairs that cover the DST changes (Fall, Spring and no change) and for freq = "H" (a.o. as it also tests for other infra-day frequencies). The test also refers to #8772 which is the issue I rephrased with a simple example in #39556.

My OP was only one case (tz=None, UTC, CET and freq=H) amongst these cases.
What was misleading in the original test is that the base frequency "H" that triggers the issue was not covered (probably because the author thought that testing with "3H" would cover "H" + other cases.

I can add a new test but I do not see what would differ from the current one (just my date_pairs would cover a full year which is not really needed for the test and my tz would be "CET" that is not covered yet other tz with DST are covered).

The small change in the test (adding freq="H" to the frequencies to test) breaks pandas before the bugfix.

I could adapt the current comment from "    # see gh-8772" to "    # see gh-8772 and gh-39556" to make it clearer ?

Comment (jreback) [reply pandas/tests/tseries/frequencies/test_inference.py:269]: sorry you are right, was misreading the fixture.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
cc @mroeschke if any comments, merge when good

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @sdementen
